packages/shell-api/src/streams.ts

```diff
@@ -0,0 +1,172 @@
+import type { Document } from 'bson';
+import { CommonErrors, MongoshInvalidInputError } from '@mongosh/errors';
+
+import { MongoshInterruptedError } from './interruptor';
+import {
+  ShellApiWithMongoClass,
+  returnsPromise,
+  shellApiClassDefault,
+} from './decorators';
+import type ShellInstanceState from './shell-instance-state';
+import StreamProcessor from './stream-processor';
+import { ADMIN_DB, asPrintable } from './enums';
+import type Mongo from './mongo';
+
+@shellApiClassDefault
+export class Streams extends ShellApiWithMongoClass {
+  public static newInstance(instanceState: ShellInstanceState) {
+    return new Proxy(new Streams(instanceState), {
+      get(target, prop) {
+        const v = (target as any)[prop];
+        if (v !== undefined) {
+          return v;
+        }
+        if (typeof prop === 'string') {
+          return target.getProcessor(prop);
+        }
+      },
+    });
+  }
+
+  constructor(private instanceState: ShellInstanceState) {
```

Comment:
Happy to change to the explicit form. And no particular reason, this is just the way I found that could get the current `Mongo`. The global `sp` instance is created inside `ShellInstanceState` so I just pass it in directly. Why go through a `Database` when we can pass in the `ShellInstanceState` directly?

Reply:
If you want to change all of these classes to take a `Mongo` instance instead of a `Database` instance in this PR, that would be fine with me, it’s semantically the “more correct thing” to do anyway. (Or otherwise, we can also try to do that separately.)

But I’d still try to keep consistent behavior here. Passing `ShellInstanceState` is not quite the same as passing a `Mongo` (or `Database`) object; consider this:

```
db = connect('mongodb://host1');
rs1 = rs;
sp1 = sp;
db = connect('mongodb://host2');
rs2 = rs; // rs2 !== rs1
sp2 = sp; // Would currently be sp2 === sp1, but should not be, one should still be able to access and use sp for host1
```

Reply:
One thing I was trying to ensure is for `sp` to always using the current connection. But I can see how that's probably not the right thing in the scenario you gave. Okay, I will change this to accept a database.